src/RADOS/RADOS.cc

```diff
@@ -0,0 +1,1356 @@
+		       std::optional<std::string_view> key)
+{
+  auto objecter = reinterpret_cast<_::RADOS*>(&impl)->objecter.get();
+  object_locator_t oloc;
+  oloc.pool = pool;
+  if (ns)
+    oloc.nspace = *ns;
+  if (key)
+    oloc.key = *key;
+
+  Objecter::LingerOp *linger_op = reinterpret_cast<Objecter::LingerOp*>(cookie);
+
+  ObjectOperation op;
+  op.watch(cookie, CEPH_OSD_WATCH_OP_UNWATCH);
+  objecter->mutate(linger_op->target.base_oid, oloc, std::move(op),
+		   {}, ceph::real_clock::now(), 0,
+		   [objecter, linger_op, c = std::move(c)]
+		   (boost::system::error_code ec, int) mutable {
+		     objecter->linger_cancel(linger_op);
+		     ceph::async::dispatch(std::move(c), ec);
+		   });
+}
+
+void RADOS::flush_watch(std::unique_ptr<VoidOpComp> c)
+{
+  auto objecter = reinterpret_cast<_::RADOS*>(&impl)->objecter.get();
+  objecter->linger_callback_flush([c = std::move(c)]() mutable {
+				    ceph::async::post(std::move(c));
+				  });
+}
+
+struct NotifyHandler {
+  boost::asio::io_context& ioc;
+  boost::asio::io_context::strand strand;
+  Objecter* objecter;
+  Objecter::LingerOp* op;
+  std::unique_ptr<RADOS::NotifyComp> c;
+
+  bool acked = false;
+  bool finished = false;
+  boost::system::error_code res;
+  bufferlist rbl;
+
+  NotifyHandler(boost::asio::io_context& ioc,
+		Objecter* objecter,
+		Objecter::LingerOp* op,
+		std::unique_ptr<RADOS::NotifyComp> c)
+    : ioc(ioc), strand(ioc), objecter(objecter), op(op), c(std::move(c)) {}
+
+  // Use bind or a lambda to pass this in.
+  void handle_ack(boost::system::error_code ec,
+		  bufferlist&&) {
+    boost::asio::post(
+      strand,
+      [this, ec]() mutable {
+	acked = true;
+	maybe_cleanup(ec);
+      });
+  }
+
+  // Notify finish callback. It can actually own the object's storage.
+
+  void operator()(boost::system::error_code ec,
+		  bufferlist&& bl) {
+    boost::asio::post(
+      strand,
+      [this, ec]() mutable {
+	finished = true;
+	maybe_cleanup(ec);
+      });
+  }
+
+  // Should be called from strand.
+  void maybe_cleanup(boost::system::error_code ec) {
+    if (!res && ec)
+      res = ec;
+    if ((acked && finished) || res)
+      boost::asio::defer(
+	ioc.get_executor(),
+	[this]() mutable {
+	  auto bl = std::move(rbl);
+	  objecter->linger_cancel(op);
+	  ceph::async::dispatch(std::move(c), res, std::move(bl));
+	});
+  }
+};
+
+void RADOS::notify(const Object& o, const IOContext& _ioc, bufferlist&& bl,
+		   std::optional<std::chrono::milliseconds> timeout,
+		   std::unique_ptr<NotifyComp> c)
+{
+  auto rados = reinterpret_cast<_::RADOS*>(&impl);
+  auto oid = reinterpret_cast<const object_t*>(&o.impl);
+  auto ioc = reinterpret_cast<const IOContextImpl*>(&_ioc.impl);
+  auto linger_op = rados->objecter->linger_register(*oid, ioc->oloc, 0);
+
+  auto x = new NotifyHandler(rados->ioctx, rados->objecter.get(), linger_op, std::move(c));
+  linger_op->on_notify_finish =
+    [cb = std::unique_ptr<NotifyHandler>(x)](boost::system::error_code ec,
```

Comment:
this NotifyHandler is shared with the `linger_notify()` callback, which can race with its destruction here. similarly, NotifyHandler is post()ing its completions to run on a strand, and something needs to extend its lifetime past the scope of this lambda

Reply:
Cherry-picked